### PR TITLE
Clarify the spec for convolution and convolution transpose when group > 1.

### PR DESCRIPTION
**Pull request by pranavsharma**
See https://github.com/onnx/onnx/issues/1282

### CODE REVIEW

File: onnx/defs/nn/defs.cc [hunk 1]
```diff
@@ -495,7 +495,9 @@ computes the output.)DOC";
         "Optionally, if dimension denotation is in effect, "
         "the operation expects the weight tensor to arrive "
         "with the dimension denotation of [FILTER_IN_CHANNEL, "
-        "FILTER_OUT_CHANNEL, FILTER_SPATIAL, FILTER_SPATIAL ...].",
+        "FILTER_OUT_CHANNEL, FILTER_SPATIAL, FILTER_SPATIAL ...]. "
```
[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnx/defs/nn/defs.cc]: sure, will fix.

File: onnx/defs/nn/defs.cc [hunk 2]
```diff
@@ -706,7 +709,9 @@ output_shape can also be explicitly specified in which case pads values are auto
         "height and width of the kernel, and M is the number "
         "of feature maps. For more than 2 dimensions, the "
         "weight shape will be (C x M x k1 x k2 x ... x kn), "
-        "where (k1 x k2 x ... x kn) is the dimension of the kernel",
+        "where (k1 x k2 x ... x kn) is the dimension of the kernel. "
```
Comment (houseroad) [inline onnx/defs/nn/defs.cc]: In convtranspose case, weight shape should be C/group x M x ...

File: onnx/defs/nn/defs.cc [hunk 3]
```diff
@@ -495,7 +495,10 @@ computes the output.)DOC";
         "Optionally, if dimension denotation is in effect, "
         "the operation expects the weight tensor to arrive "
         "with the dimension denotation of [FILTER_IN_CHANNEL, "
-        "FILTER_OUT_CHANNEL, FILTER_SPATIAL, FILTER_SPATIAL ...].",
+        "FILTER_OUT_CHANNEL/group, FILTER_SPATIAL, FILTER_SPATIAL ...]. "
+        "Note that X.shape[1] should be equal to W.shape[1] * group "
```
Comment (houseroad) [inline onnx/defs/nn/defs.cc]: It should be [FILTER_OUT_CHANNEL, FILTER_IN_CHANNEL, FILTER_SPATIAL, ...]

And FILTER_IN_CHANNEL = DATA_CHANNEL / group

and you don't need the following note I think.

File: onnx/defs/nn/defs.cc (hunk 2 above)
Comment (pranavsharma) [reply onnx/defs/nn/defs.cc]: I would actually prefer that we rename C and M. The naming is confusing. In the case of conv transpose, num_output_channels = W.shape[1]/group. W.shape[1] is actually a multiplier for num_output_channels. W.shape[0] corresponds to num_input_channels (X.shape[1]). 
If we keep the same names, the spec should be C x M/group.

File: onnx/defs/nn/defs.cc (hunk 3 above)
Comment (pranavsharma) [reply onnx/defs/nn/defs.cc]: If you look at the caffe2 implementation, X.shape[1] = W.shape[1]*group. In the terminology of the spec, this translates to DATA_CHANNEL = FILTER_OUT_CHANNEL where W.shape[1] = FILTER_OUT_CHANNEL/group.

Comment (houseroad) [reply onnx/defs/nn/defs.cc]: What's the meaning of FILTER_OUT_CHANNEL? does it mean the channel of the output?

Comment (pranavsharma) [reply onnx/defs/nn/defs.cc]: Yes, that's how the implementation interprets it. FILTER_OUT_CHANNEL is number of channels in the output.

Comment (houseroad) [reply onnx/defs/nn/defs.cc]: There are two problems in this case, 
1) the denotation filter should be [out, in, spatial, spatial]
2) the w.shape[1] * group == x.shpae[1] == C, so the W's shape should be [M, C/group, kH, kW]

File: onnx/defs/nn/defs.cc (hunk 2 above)
Comment (houseroad) [reply onnx/defs/nn/defs.cc]: Yes, this should be [C, M/group, spatial, spatial, ...]

Comment (houseroad) [reply onnx/defs/nn/defs.cc]: I would prefer to keep the names. But I am open to the new names, if you can think of any better name.

File: onnx/defs/nn/defs.cc (hunk 3 above)
Comment (pranavsharma) [reply onnx/defs/nn/defs.cc]: +1 for #2. 
Regarding #1, are you suggesting that we change the denotation as well in the spec to [FILTER_OUT_CHANNEL, FILTER_IN_CHANNEL, FILTER_SPATIAL, FILTER_SPATIAL ...] ?

File: onnx/defs/nn/defs.cc (hunk 2 above)
Comment (pranavsharma) [reply onnx/defs/nn/defs.cc]: Sure, lets keep [C, M/group, spatial, spatial, ...]

**Pull request merged**